### ISSUE TITLE
Issues with local software: Imprecise estimates, different estimates whether modelled alone or in batch, model runs considerably slow

### ISSUE THREAD

**Issue by CarloCocozza** [Pandora-IsoMemo/osteo-bior#37]
I am using the standard version of OsteoBioR 23.4.1.7 locally. I have found that estimates are often very imprecise, are different whether these are run individually or together with a number of other individuals, and that it takes considerably a lot more than using OsteoBioR online (both standard or beta versions) to generate model. 

I will now describe the steps I have followed.

-------

1) I have generated a batch model locally. In the attached files, there are 3 **archaeological** individuals. For simplicity I have labelled these 1-3, but they are, respectively: METi_239 (ind. 1 for nitrogen and ind. 2 for carbon), METi_ 237 (ind. 3 for nitrogen and ind. 4 for carbon), METi_ 223 (ind. 5 for nitrogen and ind. 6 for carbon). So in total there are 6 **modelled** individuals.
[Isotopes 1-3.xlsx](https://github.com/Pandora-IsoMemo/osteo-bior/files/13547448/Isotopes.1-3.xlsx)
[Renewal rates 1-3.xlsx](https://github.com/Pandora-IsoMemo/osteo-bior/files/13547449/Renewal.rates.1-3.xlsx)
I also attach the run saved model (with only input data and options). [Batch Model (only input data and options).zip](https://github.com/Pandora-IsoMemo/osteo-bior/files/13547469/Batch.Model.only.input.data.and.options.zip)

In this batch model, estimates for certain individuals are very imprecise and these show very high standard errors. This does not happens to all individuals. I attach a screenshot of individual 1 as reference. 
![Ind 1 bad estimates](https://github.com/Pandora-IsoMemo/osteo-bior/assets/101706974/e890fae5-9eaf-48c0-aa1f-f4b5ade02b72)

It took circa 2,30h to run this model.

2) I have then tried to model **archaeological** individual METi_239 without METi_ 237 and METi_ 223. There are still 2 **modelled** individuals as these represent, as above, nitrogen and carbon, respectively. I here attach the files for isotopes and renewal rates, plus the run saved model (with only input data and options).
[METi_239 Isotopes.xlsx](https://github.com/Pandora-IsoMemo/osteo-bior/files/13547428/METi_239.Isotopes.xlsx)
[METi_239 Renewal Rates.xlsx](https://github.com/Pandora-IsoMemo/osteo-bior/files/13547429/METi_239.Renewal.Rates.xlsx)
[Individual Model METi_239 Nitrogen and Carbon.zip](https://github.com/Pandora-IsoMemo/osteo-bior/files/13548661/Individual.Model.METi_239.Nitrogen.and.Carbon.zip)

Also here estimates are not precise for individual 1. In addition, they are different from those estimated in the batch model as above.
![Ind 1 bad estimates (Only METi_239)](https://github.com/Pandora-IsoMemo/osteo-bior/assets/101706974/5bdc3260-187c-46ae-bbde-f7ad007bae15)

It took circa 1,30h to run this model.

3) I have hence tried to use the online beta version (23.10.0) to run the same model for METi_239 as in step 2. 

The estimates are here precise. I cannot attach also the saved model due to a separate issue that I have just reported here in github. However, I attach the screenshot for the estimates for individual 1. 
![Ind 1 good estimates (Beta online)](https://github.com/Pandora-IsoMemo/osteo-bior/assets/101706974/151de55f-356b-4ec2-b6ad-051d3f8ed352)

It took circa 10 minutes to run this model.

4) In order to understand whether the issues is in the OsteoBioR version 23.4.1.7 or just in the local software, I have also tried to model the same METi_239 in the standard version online. 

Estimates are also here precise. 
![Ind 1 good estimates (standard version online)](https://github.com/Pandora-IsoMemo/osteo-bior/assets/101706974/008c0cad-9995-4d44-8304-1704309df981)

In this version I can also attach the saved run model. 
[Individual Model METi_239 Nitrogen and Carbon Standard version Local.zip](https://github.com/Pandora-IsoMemo/osteo-bior/files/13548936/Individual.Model.METi_239.Nitrogen.and.Carbon.Standard.version.Local.zip)

It took circa 3 minutes to run this model.

------------

In all steps above, I have used the following options: 5000 iterations, 1000 burn-in and 4 MCMC in all cases.

In summary, I think there is some serious issue with the local model. This generates imprecise and inaccurate estimates. I would also look into the fact that it takes a considerably enormous amount of time to run the model in comparison to the online versions.

@arunge

**Comment by arunge** [comment]
@CarloCocozza Over the weekend I testet some cases of running your model mentioned [above](https://github.com/Pandora-IsoMemo/osteo-bior/issues/37#issue-2024270351) (the 6 models version).
I added the print of the calculation time for fitting the model into the console.

| local run with docker | local run with RStudio | online run |
|---|---|---|
|![image](https://github.com/Pandora-IsoMemo/osteo-bior/assets/16759098/c6838de7-797e-4dc8-a8f7-20a29f3b5d75)|![image](https://github.com/Pandora-IsoMemo/osteo-bior/assets/16759098/35ebb725-f92d-4c1e-b4ef-298cef326773)| |
| |![image](https://github.com/Pandora-IsoMemo/osteo-bior/assets/16759098/0ac73314-3a36-482a-afc6-59d1dcbe6a78)|![image](https://github.com/Pandora-IsoMemo/osteo-bior/assets/16759098/9fd0331c-9ad4-4bf4-abfb-e4bfbc3bc9e0)|

Locally, results also had been precise, while it also happened that the online results took quite long and than also had more uncertainty.

We use absolutely the same code for all cases: online (with docker), locally with docker, local R installation. Possibly there is sometimes bad convergence.

There is a random component when using Bayesian models, and this might lead to cases where convergence is not that good if the data is few or uncertainty large.

With the new version 23.12.0 on beta the calculation time is given after the model fit. Additionally, one can now set a fixed seed in order to obtain reproducible results.

I hope this helps to understand the issue. I cannot say why all your local runs take so long while online you saw fast calculation. I also had long calculation times online.

Please try to set a seed and compare online and local runs. Maybe the issue was due to a random effect.

**Comment by arunge** [comment]
@CarloCocozza @isomemo I testet the new option to set a fixed seed locally (docker and R installation) and online. The results are exactly the same, only the calculation time differed slightly: locally with docker: 70.12 minutes, locally with R installation: 64.46 minutes, online: 64.07 minutes.

However, when using a random seed the calculation time AND the results vary a lot. Most likely there is an issue with data and randomness related convergence.

I do not see any issues regarding using the app locally or online. If you do not mind, I am closing here.

Additionaly, I will push the changes to the main app in order to make the "seed" feature also available there.

**Comment by arunge** [comment]
@CarloCocozza I have tested your model with the fixed seed "12345". The screenshots above show that it did not matter where I ran the model, it took a long time and results where less precise on all systems.

Different seeds gave very different results for the data [above](https://github.com/Pandora-IsoMemo/osteo-bior/issues/37#issue-2024270351). I do not believe that there is an issue with your docker, the code is the same on all systems and thus results _must_ be the same when we use the same _seed_.
